### PR TITLE
Add type hints to lakeFS file and file system [ENG-123]

### PR DESCRIPTION
**Pull request by nicholasjng**
These type hints are the strictest they can be without excessive errors.

This leads to a few relaxations, though:

* The `outfile` argument was reduced to `typing.Any`, since it is optionally assigned to a str or IO class, which is hard to capture and model correctly.
* Paths are currently limited to `str`s, since they are fed directly to the lakeFS URI regex parser.

Followup PRs could e.g. add support for `PathLike`s as {r,l}paths.

### CODE REVIEW

**Review by AdrianoKF — COMMENTED** [review]
https://github.com/appliedAI-Initiative/lakefs-spec/blob/f50c58f23fa5831c23546c1a2489a9d5bd40797b/src/lakefs_spec/config.py#L17

The return could be typed as `Self` (`from typing_annotations import Self`), see [PEP 673](https://peps.python.org/pep-0673/).

(also looking at the same method, the error message for missing `PyYAML` is slightly inaccurate: it says that the config file exists, even when it might not)

**Review by AdrianoKF — COMMENTED** [review]
https://github.com/appliedAI-Initiative/lakefs-spec/blob/f50c58f23fa5831c23546c1a2489a9d5bd40797b/src/lakefs_spec/errors.py#L16

Remove `# type: ignore` here (doesn't seem to be required in any case)

**Comment by nicholasjng** [comment]
> (also looking at the same method, the error message for missing `PyYAML` is slightly inaccurate: it says that the config file exists, even when it might not)

Hm, good point. We do in fact only call `LakectlConfig.read()` if the `path` input exists, but to make this portable, the existence check could probably be carried out again in the method?

But this raises the question of how to proceed if `path.exists() is False`. Return an empty object and risk silent errors, or raise?

**Review by AdrianoKF — COMMENTED** [review]
https://github.com/appliedAI-Initiative/lakefs-spec/blob/f50c58f23fa5831c23546c1a2489a9d5bd40797b/src/lakefs_spec/spec.py#L180

This is untyped

File: src/lakefs_spec/spec.py [hunk 1]
```diff
@@ -386,7 +386,7 @@ def info(self, path: str, **kwargs: Any) -> dict[str, Any]:
         self.run_hook(FSEvent.INFO, HookContext.new(path))
         return statobj
 
-    def ls(self, path, detail=True, **kwargs):
+    def ls(self, path: str, detail: bool = True, **kwargs: Any) -> list[Any]:
```
Comment (AdrianoKF) [inline src/lakefs_spec/spec.py:389]: I'd be fine with `list`, since it is equivalent to `list[Any]`

**Comment by AdrianoKF** [comment]
> > (also looking at the same method, the error message for missing `PyYAML` is slightly inaccurate: it says that the config file exists, even when it might not)
> 
> Hm, good point. We do in fact only call `LakectlConfig.read()` if the `path` input exists (see the file system constructor), but to make this portable, the existence check could probably be carried out again in the method?
> 
> But this raises the question of how to proceed if `path.exists() is False`. Return an empty object and risk silent errors, or raise?

For now, we can get away by (a) rewording the error message, and/or (b) reading the file first and then importing `PyYAML` (this would eliminate a race condition between the existence check and trying to read the file, while still keeping the error message relevant).

**Comment by AdrianoKF** [comment]
Strong agree that we should allow passing `str | os.PathLike[str] | Path` in a follow-up issue (see the [internal handling](https://github.com/fsspec/filesystem_spec/blob/096e9a186ffdb39b26b6afdf38c610d092c85f95/fsspec/utils.py#L327) in `fsspec`.

**Comment by nicholasjng** [comment]
> https://github.com/appliedAI-Initiative/lakefs-spec/blob/f50c58f23fa5831c23546c1a2489a9d5bd40797b/src/lakefs_spec/config.py#L17
> 
> The return could be typed as `Self` (`from typing_annotations import Self`), see [PEP 673](https://peps.python.org/pep-0673/).

How do we do this for Python >=3.11, where it's built in? I would prefer spinning off a follow-up for this, since it would also carry an optional dependency on `typing_extensions`.

**Comment by AdrianoKF** [comment]
> > https://github.com/appliedAI-Initiative/lakefs-spec/blob/f50c58f23fa5831c23546c1a2489a9d5bd40797b/src/lakefs_spec/config.py#L17
> > 
> > The return could be typed as `Self` (`from typing_annotations import Self`), see [PEP 673](https://peps.python.org/pep-0673/).
> 
> How do we do this for Python >=3.11, where it's built in? I would prefer spinning off a follow-up for this, since it would also carry an optional dependency on `typing_extensions`.

AFAIK, `typing_extensions` correctly uses the language builtins for versions that support them and only adds backports for older versions, so no changes would be required. However, I'm okay with moving this into a follow-up issue.

**Comment by nicholasjng** [comment]
> AFAIK, `typing_extensions` correctly uses the language builtins for versions that support them (edit: [it does](https://github.com/python/typing_extensions/blob/9de9fd613913faee5db317f827f5eec3755d8a92/src/typing_extensions.py#L1795-L1813)) and only adds backports for older versions, so no changes would be required. However, I'm okay with moving this into a follow-up issue.

So by "no changes", you mean `import typing_extensions` regardless of Python version? That would mean taking up `typing_extensions` as an unconditional dependency. Is this worth it for these few classmethod calls we have?

**Comment by AdrianoKF** [comment]
> So by "no changes", you mean `import typing_extensions` regardless of Python version? That would mean taking up `typing_extensions` as an unconditional dependency. Is this worth it for these few classmethod calls we have?

Exactly. If it's just about `Self`, we can also ignore it for now (as the current typing correctly captures everything - there's a few other edge cases that only `Self` addresses, but we don't have them here).

File: src/lakefs_spec/spec.py (hunk 1 above)
Comment (nicholasjng) [reply src/lakefs_spec/spec.py]: Changed in [cc14128](https://github.com/appliedAI-Initiative/lakefs-spec/pull/123/commits/cc141285497fdb326d1aa4e3aae20edbe28ce25e).

**Comment by nicholasjng** [comment]
> https://github.com/appliedAI-Initiative/lakefs-spec/blob/f50c58f23fa5831c23546c1a2489a9d5bd40797b/src/lakefs_spec/errors.py#L16
> 
> Remove `# type: ignore` here (doesn't seem to be required in any case)

Removed in [cc14128](https://github.com/appliedAI-Initiative/lakefs-spec/pull/123/commits/cc141285497fdb326d1aa4e3aae20edbe28ce25e).

File: src/lakefs_spec/spec.py (hunk 1 above)
Comment (AdrianoKF) [reply src/lakefs_spec/spec.py]: No, it's not ;)

Comment (nicholasjng) [reply src/lakefs_spec/spec.py]: Sorry, meant [cc14128](https://github.com/appliedAI-Initiative/lakefs-spec/pull/123/commits/cc141285497fdb326d1aa4e3aae20edbe28ce25e).

**Review by AdrianoKF — APPROVED** [review]

**Pull request merged**